src/FieldWarning/Assets/Terrain/Scripts/TerrainMap.cs

```diff
@@ -34,67 +36,226 @@ public class TerrainMap
+
+    }
+
+    public void WriteHeightMap(string path)
+    {
+        
         int nEntry = 2 * mapSize + 2 * EXTENSION;
-        map = new byte[nEntry, nEntry];
+        BinaryWriter writer = new BinaryWriter(File.Open(path, FileMode.Create, FileAccess.Write, FileShare.Read));
+        for (int x = 0; x < nEntry; x++)
+        {
+            var first = true;
+            float temp = 0;
+            float last = 0;
+            int lastcnt = 0;
 
-        WaterBasic water = (WaterBasic)GameObject.FindObjectOfType(typeof(WaterBasic));
-        waterHeight = water.transform.position.y;
-        for (int x = 0; x < nEntry; x++) {
             for (int z = 0; z < nEntry; z++)
-                map[x, z] = (byte)(terrain.SampleHeight(PositionOf(x, z)) > waterHeight ? PLAIN : WATER);
+            {
+
+
+                temp = terrain.SampleHeight(PositionOf(x, z));
+
+                //map[x, z] = (byte)(terrain.SampleHeight(PositionOf(x, z)) > waterHeight ? PLAIN : WATER);
+
+                if (last == temp || first)
+                {
+
+                    lastcnt++;
+                    first = false;
+
+                }
+                else
+                {
+                    writer.Write(temp);
+                    writer.Write(lastcnt);
+
+
+                    lastcnt = 0;
+                    first = true;
+
+                }
+
+                last = temp;
+
+
+            }
+            // if there are multiple of same type on last iteration.. dump to file
+            if (!first)
+            {
+                writer.Write(temp);
+                writer.Write(lastcnt);
+
+            }
+
+            writer.Write((int)'\n');
+
         }
 
-        foreach (TreeInstance tree in terrain.terrainData.treeInstances) {
-            Vector3 treePosition = Vector3.Scale(tree.position, terrain.terrainData.size) + terrain.transform.position;
-            AssignCircularPatch(treePosition, TREE_RADIUS, FOREST);
+        writer.Close();
+    }
+
+    public void ReadHeightMap(string path)
+    {
+        //TODO : not much error checking is done in thsi function
+
+        int nEntry = 2 * mapSize + 2 * EXTENSION;
+        var file = File.ReadAllBytes(path);
+
+        //var last_notify_msec = 0;
+        int x = 0;
+        int z = 0;
+        MemoryStream ms = new MemoryStream(file);
+        BinaryReader reader = new BinaryReader(ms);
+
+        while (reader.BaseStream.Position != reader.BaseStream.Length)
+        {
+
+            // <type><-><amount><,>[\n]
+            byte[] type = reader.ReadBytes(4);
+
+            if (BitConverter.ToInt32(type, 0) == (int)0x0a)
+            {
+                z = 0;
+                x++;
+            }
+            else
+            {
+
+                // since we already read a byte but we need 4 bytes
+                int len = reader.ReadInt32();
+
+                var bType = (byte)(BitConverter.ToSingle(type, 0) > waterHeight ? PLAIN : WATER);
+
+                var zEnd = z + len;
```

Comment:
I feel like the names could be a lot more descriptive than `z` and `x` and `len`. Is it `fileLength`? Is `z` a coordinate? Is `zEnd` the highest allowed position on the z axis?

Also `var` is a bit overused, it's fine if the type is obvious but in the line `var zEnd = z + len;` this is not the case.